### DISCUSSION TITLE
How to retrieve widgets across config reloads?

### DISCUSSION THREAD

**Discussion by holocronweaver** [qtile/qtile#5039] [Q&A] [answered]
I have a `@lazy.function` which retrieves the groupbox widget via `widgets_map` like this:
```
@lazy.function
def show_only_favorite_groups(qtile, favorites: list[str]):
  qtile.widgets_map['groupbox'].visible_groups = favorites
  qtile.widgets_map['groupbox'].draw()
```
However after I reload my config using `lazy.reload_config`, the original `groupbox` has been replaced by `groupbox_1`, and so the `widgets_map` call above will return a discarded widget instead of the current one.

Is this a bug - should the original widgets have been discarded and replaced on reload with the same names as before the reload? If not clearing the widget map on reload is intended behavior, is there a way to retrieve widgets that is robust across config reloads? 

I'm using qtile 0.28.1.

Thanks!

**Comment by ervinpopescu** [comment]
Either set a custom `name` when defining the widget or get it from the bar by matching on the `groupbox` name like so:
```py
@lazy.function
def show_only_favorite_groups(qtile, favorites: list[str]):
    groupbox_name = next(w for w in qtile.widgets_map if "groupbox" in w)
    groupbox = qtile.widgets_map[groupbox_name]
    groupbox.visible_groups = favorites
    logger.info(groupbox.visible_groups)
    groupbox.draw()
    groupbox.bar.draw()
```

> Is this a bug - should the original widgets have been discarded and replaced on reload with the same names as before the reload?

If you check your logs, you see a similar log message to the one below:
```
2024-10-18 13:44:20,904 INFO libqtile bar.py:_configure():L354 The following widgets were renamed in qtile.widgets_map: qtemirror_1, qtemirror_2, qtemirror_3, qtemirror_4, qtemirror_5, qtemirror_6, qtemirror_7 To bind commands, rename the widget or use lazy.widget[new_name].
```

So, it's expected behavior. Qtile does not guarantee same naming, although I think based on the code that renames the widgets, there is no random naming involved, which should mean you will see the same naming based on the widget's index in the widget list.

> If not clearing the widget map on reload is intended behavior, is there a way to retrieve widgets that is robust across config reloads?

I'm not sure what you mean, the widgets_map **will** be discarded/cleared on reload, how else would qtile reload the config?

**Comment by elParaguayo** [comment]
> Is this a bug?

Yes. That absolutely should not be happening.

**Reply by ervinpopescu** [reply]
> So, it's expected behavior. Qtile does not guarantee same naming, although I think based on the code that renames the widgets, there is no random naming involved, which should mean you will see the same naming based on the widget's index in the widget list.

But was I wrong tho? Or do you mean if you have only one `groupbox`, it shouldn't turn into `groupbox_1`?

**Reply by elParaguayo** [reply]
Correct. If there's only one `groupbox` it should never become `groupbox_1`. Those suffixes are added when there are multiple widgets with the same name running at the same time.

**Reply by ervinpopescu** [reply]
Oh, okay, I wasn't sure about that, but I don't think that's happening here, tbh, because I haven't seen it in my config. I know, it's a bit condescending to say this, but it really is an "easy" algorithm which I don't see creating edge cases or stuff like that. 

See my widget names below:
```bash
$ qticc cmd-obj -o bar bottom -f items -a widget | jq '.[1][]' -r | sed -E 's/qtemirror/sm_sp/'
arch_logo
sm_sp
groupbox
sm_sp
currentlayouticon
sm_sp
tasklist
sm_sp
chord
sm_sp
systray
sm_sp
githubnotifications
sm_sp
battery
battery_icon
battery_sep
bt_battery
sm_sp
mouse_over_clock
sm_sp
wallpaper
powermenu
```

With spacer mirrors:
```bash
$ qticc cmd-obj -o bar bottom -f items -a widget | jq '.[1][]' -r
arch_logo
sm_sp
groupbox
qtemirror
currentlayouticon
qtemirror
tasklist
qtemirror
chord
qtemirror
systray
qtemirror
githubnotifications
qtemirror
battery
battery_icon
battery_sep
bt_battery
qtemirror
mouse_over_clock
qtemirror
wallpaper
powermenu
```

`widgets_map`
```bash
$ qticc cmd-obj -f eval -a 'self.widgets_map' | jq '.[1]' -r | sed "s/</'</g;s/>/>'/g" | qtile_to_json
```
```json
{
  "arch_logo": "<qtile_extras.widget.image.Image object at 0x78322fe37f20>",
  "sm_sp": "<libqtile.widget.spacer.Spacer object at 0x783238e0f7a0>",
  "groupbox": "<libqtile.widget.groupbox.GroupBox object at 0x783238e42960>",
  "qtemirror": "<qtile_extras.widget.mirror.QTEMirror object at 0x78322f4eecf0>",
  "currentlayouticon": "<qtile_extras.widget.currentlayout.CurrentLayoutIcon object at 0x78322ddd7380>",
  "qtemirror_1": "<qtile_extras.widget.mirror.QTEMirror object at 0x78322fe56ba0>",
  "tasklist": "<libqtile.widget.tasklist.TaskList object at 0x783238fbea50>",
  "qtemirror_2": "<qtile_extras.widget.mirror.QTEMirror object at 0x78322e847260>",
  "chord": "<libqtile.widget.chord.Chord object at 0x78322ea0c860>",
  "qtemirror_3": "<qtile_extras.widget.mirror.QTEMirror object at 0x783238ff3cb0>",
  "systray": "<extras.widgets.systray.Systray object at 0x78322ea051f0>",
  "qtemirror_4": "<qtile_extras.widget.mirror.QTEMirror object at 0x783238fdfef0>",
  "githubnotifications": "<qtile_extras.widget.githubnotifications.GithubNotifications object at 0x78322ec4d3a0>",
  "qtemirror_5": "<qtile_extras.widget.mirror.QTEMirror object at 0x783246bf6b70>",
  "battery": "<extras.widgets.battery.Battery object at 0x78322e98da60>",
  "battery_icon": "<extras.widgets.battery_icon.BatteryIcon object at 0x783239aa28a0>",
  "battery_sep": "<libqtile.widget.sep.Sep object at 0x78322ec9b380>",
  "bt_battery": "<extras.widgets.bt_bat.BtBattery object at 0x783238e46690>",
  "qtemirror_6": "<qtile_extras.widget.mirror.QTEMirror object at 0x783239a06000>",
  "mouse_over_clock": "<extras.widgets.mouseoverclock.MouseOverClock object at 0x783238ff7ce0>",
  "qtemirror_7": "<qtile_extras.widget.mirror.QTEMirror object at 0x783238e44e00>",
  "wallpaper": "<libqtile.widget.textbox.TextBox object at 0x78322e9d0380>",
  "powermenu": "<libqtile.widget.textbox.TextBox object at 0x783246bdc7d0>"
}
```

> Those suffixes are added when there are multiple widgets with the same name running at the same time.

In the same bar* might I add :)

Edit: although... why do we have multiple bars but only one widgets_map? Does the `widgets_map` concatenate all the widgets from all bars?

**Reply by elParaguayo** [reply]
I also haven't seen this behaviour so I suspect there's something else going on.

> In the same bar might I add :)

No, the naming scheme ignores bars. You should get the renaming if you have a groupbox in each bar, one would be `groupbox` and the other would be `groupbox_1` (unless you're using the same widget instance, in which case the second would be a mirror).

**Reply by ervinpopescu** [reply]
Okay, that makes sense. And how about this:
> Edit: although... why do we have multiple bars but only one widgets_map? Does the widgets_map concatenate all the widgets from all bars?

**Reply by elParaguayo** [reply]
Yes - the map collates all widgets into a single dict.

**Comment by elParaguayo** [comment]
@holocronweaver I can't recreate your issues. Here's the output from a `qtile shell` session:
```
 > eval(list(self.widgets_map.keys()))
(True,
 "['currentlayout', 'groupbox', 'prompt', 'windowname', 'chord', 'default', "
 "'textbox', 'systray', 'clock', 'quickexit']")
 > reload_config()
 > eval(list(self.widgets_map.keys()))
(True,
 "['currentlayout', 'groupbox', 'prompt', 'windowname', 'chord', 'default', "
 "'textbox', 'systray', 'clock', 'quickexit']")
```
As you can see, widget names are unchanged after a config reload.

Can you run the same commands and post the output here?

NB the above output is with the default config so please try using that too to see if you get the same problem.

**Reply by holocronweaver** [reply]
Turns out that when the `Notify` widget fails to load properly (in my case because `dbus-next` was missing on a fresh OS install), like so:
```
2024-10-06 20:14:06,597 WARNING libqtile notify.py:_configure():L97 You must install dbus-next to use the Notify widget.
```
then something goes awry with the construction of `widgets_map` during reloads and new duplicates occur every config reload (resulting in names with `_1`, `_2`, etc., incrementing with each reload). When the `Notify` widget issue was resolved by installing `dbus-next`, the duplication stopped.

I am not sure if this is expected behavior when a widget fails to load.

BTW, thanks to @elParaguayo and @ervinpopescu for looking into this!

**Reply by elParaguayo** [reply, marked as answer]
Great detective work! That's definitely a bug if that's what is happening.

EDIT: Confirmed - I can replicate this.

EDIT 2: And I've got a fix for it - #5051
